### PR TITLE
update Uptycs jwt token authentication and add new commands

### PR DESCRIPTION
**Pull request by cgadde-uptycs**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
1) Updated the jwt token authentication module as per latest Uptycs public api
2) Added new commands to enhance user experience

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by RotemAmit** [comment]
> Thank you for your contribution. Your generosity and caring are unrivaled! Make sure to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form, so our content wizard @RotemAmit will know the proposed changes are ready to be reviewed.

Thank you for your contribution! Please register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form. Thank you!

**Comment by RotemAmit** [comment]
> > Thank you for your contribution. Your generosity and caring are unrivaled! Make sure to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form, so our content wizard @RotemAmit will know the proposed changes are ready to be reviewed.
> 
> Thank you for your contribution! Please register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form. Thank you!

Thank you for your contribution!
This is a kindly reminder to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form. Thank you again.

File: Packs/Uptycs/Integrations/Uptycs/README.md [hunk 1]
```diff
@@ -2578,6 +2592,698 @@ There is no context output for this command.
...
+---
+Get an Uptycs lookup table details
+##### Base Command
+
+`uptycs-get-lookuptable`
+##### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| table_id | id of the table | Required |
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/README.md:3138]:
```suggestion
| table_id | table id | Required |
```

File: Packs/Uptycs/Integrations/Uptycs/README.md [hunk 2]
```diff
@@ -2578,6 +2592,698 @@ There is no context output for this command.
...
+```
+
+##### Human Readable Output
+### Uptycs look up tables
+|active|createdAt|createdBy|customerId|dataLookupTable|description|fetchRowsquery|forRuleEngine|id|idField|name|rowCount|seedId|updatedAt|updatedBy|
+|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
+|TRUE|2023-04-21T08:27:20.888Z|f976bda8-d5dc-468f-8283-20d5368352e2|b1c3b08c-eedd-4b94-8ba0-9ca322401016|null|null|SELECT id_field_value,data FROM upt_lookup_rows WHERE lookup_table_id = '3fdb051b-b38b-4792-9daa-0a88ba4fae53'|TRUE|3fdb051b-b38b-4792-9daa-0a88ba4fae53|remote_address|Test 1|24|null|2023-04-25T04:11:04.664Z|f976bda8-d5dc-468f-8283-20d5368352e2|
+
+
+### 44. uptycs-get-lookuptables
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/README.md:3179]:
```suggestion
### 45. uptycs-get-lookuptables
```

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 3]
```diff
@@ -1798,7 +1798,539 @@ script:
...
+    arguments:
+    - name: asset_id
+      required: true
+      description: asset id
+      isArray: false
+      secret: false
+      default: false
+    - name: tag_id
+      required: true
+      description: tag id
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.yml:2312]:
```suggestion
      description: Disassociate the tag matching this id with the given asset_id
```

File: Packs/Uptycs/Integrations/Uptycs/README.md [hunk 4]
```diff
@@ -2578,6 +2592,698 @@ There is no context output for this command.
...
+Disassociate an asset with a tag
+##### Base Command
+
+`uptycs-delete-assets-tag`
+##### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| tag_id | disassociate the tag matching this id | Required |
+| asset_id | disassociate the asset matching this asset id | Required |
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/README.md:3268]:
```suggestion
| tag_id | Disassociate the tag matching this id with the given asset_id | Required |
| asset_id | Disassociate the asset matching this asset id with the given tag_id | Required |
```

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 5]
```diff
@@ -1798,7 +1798,539 @@ script:
...
+      secret: false
+      default: false
+  - name: uptycs-delete-assets-tag
+    description: disassociate an asset with a tag
+    deprecated: false
+    execution: false
+    arguments:
+    - name: asset_id
+      required: true
+      description: asset id
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.yml:2306]:
```suggestion
      description: Disassociate the asset matching this asset id with the given tag_id
```

**Review by RotemAmit — COMMENTED** [review]
Thank you for your contribution! Nice work!
A few comments:
* Please make sure that the description of the commands and arguments in the YML matches the README.
* Please add unit tests for the new commands. You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.

**Review by RotemAmit — COMMENTED** [review]
Thank you for your contribution! Nice work!
A few comments:
* Please make sure that the description of the commands and arguments in the YML matches the README.
* Please add unit tests for the new commands. You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.

**Comment by cgadde-uptycs** [comment]
@RotemAmit updated all description as per your comments. Thank you.

**Review by RotemAmit — COMMENTED** [review]
Nice work!
* Please add unit tests for the new commands. You can use the following guide - https://xsoar.pan.dev/docs/integrations/unit-testing as well as examples of unit tests in the HelloWorld integration and other packs from the content repository.
* Please make sure that the description of the Context Output in the README and YML files matches.
If you have any questions feel free to write me here or via DFIR.

**Comment by cgadde-uptycs** [comment]
@RotemAmit  I synced all description in yml and Readme. Added a testcase, Can you please check and let me know if this is fine ?  If this is fine, I can add remaining command testcases.

**Comment by RotemAmit** [comment]
> @RotemAmit I synced all description in yml and Readme. Added a testcase, Can you please check and let me know if this is fine ? If this is fine, I can add remaining command testcases.

Hi, overall the test looks ok, but it fails when running pytest. You can find the error [here](https://app.circleci.com/pipelines/github/demisto/content/272686/workflows/4db714ce-bdbc-4193-91c6-72c27cae4b05/jobs/706410), please take a look and fix it. Moreover, there is a Flake8 error that you can see in the same link. Please fix it too.

**Comment by cgadde-uptycs** [comment]
@RotemAmit Need you help. mocker used in tests is unable to set mandatory parameters domain and customer_id, Hence testcases are failing. 2 other parameters key and secret were getting set.

**Comment by RotemAmit** [comment]
> @RotemAmit Need you help. mocker used in tests is unable to set mandatory parameters domain and customer_id, Hence testcases are failing. 2 other parameters key and secret were getting set.

You can try to mock all the global variables inside the tests as well, like in the following example: 
mocker.patch("Uptycs.CUSTOMER_ID", new=CUSTOMER_ID_VALUE)

**Comment by cgadde-uptycs** [comment]
@RotemAmit Completed adding testcases. Please check.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 6]
```diff
@@ -2199,17 +2249,131 @@ def uptycs_post_saved_query_command():
...
 
-    if query_results == 0:
-        return False
-    else:
+    if query_results.get('items') is not None:
+        if len(query_results.get('items')) == 0:
+            return False
         return True
+    else:
+        return False
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:2376]: If a command fail (in this case the test_module), it should raise an exception with an informative message with the failing reason. Please update this.

File: Packs/Uptycs/ReleaseNotes/1_0_9.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+
+#### Integrations
+##### Uptycs
+- updated Uptycs jwt token authentication and added new commands.
```
Comment (RotemAmit) [inline Packs/Uptycs/ReleaseNotes/1_0_9.md:4]: Please add a list of the new commands, as in the example in the screenshot.
image

Comment (RotemAmit) [inline Packs/Uptycs/ReleaseNotes/1_0_9.md:4]:
```suggestion
- updated Uptycs jwt token authentication and added new commands.
- Updated the Docker image to: *demisto/uptycs:1.0.0.63766*.
```
Please also add a line on the updated docker image.

**Review by RotemAmit — COMMENTED** [review]
Nice work!
* Please add unit tests to the commands uptycs-post-lookuptable-data and to the command uptycs-create-lookuptable .
* Let's schedule a demo in order make sure that everything works as expected. I am sending you an email.

File: Packs/Uptycs/ReleaseNotes/1_0_9.md (hunk 7 above)
Comment (cgadde-uptycs) [reply Packs/Uptycs/ReleaseNotes/1_0_9.md]: updated

Comment (cgadde-uptycs) [reply Packs/Uptycs/ReleaseNotes/1_0_9.md]: updated

**Comment by cgadde-uptycs** [comment]
@RotemAmit Main Uptycs library was unable to pick file from test_data, Can you please tell me how to handle it ?
image

**Comment by RotemAmit** [comment]
> @RotemAmit Main Uptycs library was unable to pick file from test_data, Can you please tell me how to handle it ? image

Hi,
You need to mock demisto.getFilePath, just like you mocked demisto.args. It's return value should be like this: {'id': id, 'path': 'test/test.txt', 'name': 'test.txt'}.

**Comment by cgadde-uptycs** [comment]
@RotemAmit Thank you for the help. Completed adding testcase for the remaining 2 commands.

**Comment by cgadde-uptycs** [comment]
@RotemAmit As discussed during the demo. I made the following changes.

1) For command "uptycs-edit-lookuptable" , argument "active" was not working. Fixed it :
Screenshot from 2023-07-03 17-32-15

2) Added error message for 403 response for command 'uptycs-get-carves-download-file' :
Screenshot from 2023-07-03 17-29-11

3) For command uptycs-get-carves , optional argument path was not working. Fixed it : 
Screenshot from 2023-07-03 17-25-36

**Review by RotemAmit — COMMENTED** [review]
Nice work!
Just a few last comments.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 8]
```diff
@@ -1798,7 +1798,544 @@ script:
...
+    execution: false
+    name: uptycs-get-carves-download-file
+  - name: uptycs-get-asset-with-id
+    deprecated: false
+    description: Get an Uptycs asset details.
+    execution: false
+    arguments:
+    - name: asset_id
+      required: true
+      description: Uptycs asset id
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: description updated.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 9]
```diff
@@ -1798,7 +1798,544 @@ script:
...
+      description: Asset live status
+      type: boolean
+  - name: uptycs-get-tag
+    description: Return Uptycs asset tag details.
+    deprecated: false
+    execution: false
+    arguments:
+    - name: tag_id
+      required: true
+      description: Uptycs tag id
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: description updated.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 10]
```diff
@@ -1798,7 +1798,544 @@ script:
...
+    arguments:
+    - name: name
+      required: true
+      description: Name of a lookup table.
+      isArray: false
+      secret: false
+      default: false
+    - name: id_field
+      required: true
+      description: Id field for the table
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: description updated.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 11]
```diff
@@ -1798,7 +1798,544 @@ script:
...
+      description: Query used to fetch rows
+      type: string
+  - name: uptycs-get-lookuptable
+    description: Get an Uptycs lookup table details.
+    deprecated: false
+    execution: false
+    arguments:
+    - name: table_id
+      required: true
+      description: Id of table in Uptycs DB
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: description updated.

**Comment by cgadde-uptycs** [comment]
@RotemAmit descriptions updated as suggested. Thanks for the inputs.

File: Packs/Uptycs/Integrations/Uptycs/README.md [hunk 12]
```diff
@@ -2578,6 +2592,726 @@ There is no context output for this command.
...
+---
+Return Uptycs asset tag details
+##### Base Command
+
+`uptycs-get-tag`
+##### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| tag_id | Uptycs tag id. Get id from list of tags with command uptycs-get-tags. | Required |
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/README.md:2821]:
```suggestion
| tag_id | Uptycs tag id. Get the id from the list of tags with the command uptycs-get-tags. | Required |
```
Please update this change in all of the relevant places in the yml and README.

File: Packs/Uptycs/Integrations/Uptycs/README.md [hunk 13]
```diff
@@ -2578,6 +2592,726 @@ There is no context output for this command.
...
+---
+Update csv data for a look up table.
+##### Base Command
+
+`uptycs-post-lookuptable-data`
+##### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| table_id | Look up table id. Get id from list of tables with command uptycs-get-lookuptables.| Required |
```
Comment (RotemAmit) [inline Packs/Uptycs/Integrations/Uptycs/README.md:3054]:
```suggestion
| table_id | Look up table id. Get the id from the list of tables with the command uptycs-get-lookuptables.| Required |
```
Please update this change in all of the relevant places in the yml and README.

**Review by RotemAmit — COMMENTED** [review]
Nice work!

**Comment by cgadde-uptycs** [comment]
@RotemAmit updated descriptions as per comments. Thank you.

**Review by RotemAmit — APPROVED** [review]

**Pull request merged**